### PR TITLE
Observer pattern

### PR DESCRIPTION
**Pull request by qnislavnachev**
there is an implementation of observer pattern, it`s about stocks in a shop, when some stock is added or sold, observer  updating the information about this stock . PS : I will add some tests.

### CODE REVIEW

File: src/observer/Item.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package observer;
+
+public interface Item {
```
Comment (tsony-tsonev) [inline src/observer/Item.java]: very very bad name for interface, please read the following books that are shared with you on google drive: `Clean Code` and `Clean Coder`

File: src/observer/Shop.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    /**
+     * sell stock from the shop if it available and add it
+     * to sold stock list
+     */
+    public void sell(Stock stock) {
+        if (availableStock.contains(stock)) {
+            availableStock.remove(stock);
+            soldStock.add(stock);
+            stock.notifyObservers("was sold !");
```
Comment (tsony-tsonev) [inline src/observer/Shop.java]: what is the idea behind removing the Stock then notifying it that something was sold ???

File: src/observer/Item.java (hunk 1 above)
Comment (qnislavnachev) [inline src/observer/Item.java]: ok i will

Comment (qnislavnachev) [inline src/observer/Item.java]: fixed

File: src/observer/Shop.java (hunk 2 above)
Comment (qnislavnachev) [inline src/observer/Shop.java]: fixed

Comment (tsony-tsonev) [inline src/observer/Shop.java]: I asked what is the idea not to remove it, actually the whole idea is wrong. You either miss-understood the pattern or the task description. You should have a Shop and two classes one managing availble stock the other the manages the sold stock. You pass this classes to the shop, the shop registers them as observers. When you sell or add some stock you notify the observers when they see what was the change they change their lists of stocks.

Comment (qnislavnachev) [inline src/observer/Shop.java]: Oh yeah I see task description now, thats my fault I am sorry

**Comment by tsony-tsonev** [comment]
Please explain me how you understand the task, cause it is still wrong, as I sad before the classes managing the products should be observers inside the shop. And yet again you keep observers in the products????

**In the reply to this comment I want you to describe me with your own words what you think should be done, what should be the workflow of the task and by what logic.**

**Comment by tsony-tsonev** [comment]
Also I want UML sequence diagram of the workflow you think the program should have.

**Comment by qnislavnachev** [comment]
Well I think we should have 1 class for creating product, 2 classes that manage shop sales and product availability. When we create shop object it should create all needed observers for the shop, after that we have 2 methods add and sell, when we add some product it register product to the right observer, when we use sell method, product should be register to sales observer, if we have 5-6 observers and when we add some product it should register it to the right observers, that`s my opinion.

**Comment by tsony-tsonev** [comment]
The idea of the observer pattern is to tell all the observers for a change. The listeners for this change determine what to do with the change one their own. In your case you can pass an action type along with the stock so you can determine what to do with the changed stock in the observer. The stock itself is not observer, if you go to bicycle store and by one of the bicycles who should be notified about the purchase?.....  the rest of the bicycles or you would want to notify the shop manager?..... In your case the you notify the rest of the stock instead of the manager. The manager itself (the observer) to you are the two classes managing the available stock and the sold stock.

**Comment by tsony-tsonev** [comment]
I still wont the sequence diagram

**Comment by qnislavnachev** [comment]
I sent you on a email

**Comment by tsony-tsonev** [comment]
Cause the task is still wrong

**Comment by tsony-tsonev** [comment]
Attatch it here in a comment or in the PR description

**Comment by tsony-tsonev** [comment]
We use http://plantuml.com/ for our sequence digram it has simple sintaxis. Please recreate your diagram with plant uml so I can understand it correctly.

**Comment by qnislavnachev** [comment]
diagram here is the diagram

**Comment by tsony-tsonev** [comment]
bbb

this is the correct workfllow of the pattern

**Comment by tsony-tsonev** [comment]
LGTM

File: test/testobserver/ObserverTest.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
+package testobserver;
+
+import observer.*;
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+public class ObserverTest {
```
Comment (mgenov) [inline test/testobserver/ObserverTest.java]: You are testing Shop, not the Observer. 

ShopTest

File: src/observer/Observer.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package observer;
+
+public interface Observer {
+
+    void update(Product product, ActionType type);
+
+    void statistics();
```
Comment (mgenov) [inline src/observer/Observer.java]: statistics is an action not an event, so this method should not be part of this interface.

File: test/testobserver/ObserverTest.java (hunk 3 above)
Comment (qnislavnachev) [inline test/testobserver/ObserverTest.java]: Renamed to ShopTest.

File: src/observer/Observer.java (hunk 4 above)
Comment (qnislavnachev) [inline src/observer/Observer.java]: no longer statistic method belong to observer interface.

Comment (qnislavnachev) [inline src/observer/Observer.java]: Can I merge that pull request or I have to make some changes on it ?

**Pull request merged**